### PR TITLE
feat: add driver and changes in document needed for sparse pipeline

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

File: jina/drivers/encode.py [hunk 1]
```diff
@@ -34,3 +34,22 @@ def _apply_all(self, docs: 'DocumentSet', *args, **kwargs) -> None:
                 raise LengthMismatchException(msg)
             for doc, embedding in zip(docs_pts, embeds):
                 doc.embedding = embedding
+
+
+class ScipySparseEncodeDriver(FlatRecursiveMixin, BaseEncodeDriver):
```
Comment (JoanFM) [inline jina/drivers/encode.py:39]: is there a unified way to work with this with other sparse vectors? Isn't it better to detect the type in `EncodeDriver`? There may be no need for an extra `encodedriver`?

Comment (davidbp) [reply jina/drivers/encode.py:39]: We could do this in EncodeDriver adding a generator that allow us to iterate over the rows of different 'tensor types' with the same unified interface:

```

            for doc, embedding in zip(docs_pts, embeds):
                doc.embedding = embedding
```
by
```
            for doc, embedding in zip(docs_pts, matrix_iterator(embeds)):
                doc.embedding = embedding
```

where 

```
def matrix_iterator(X):
    n_rows = X.shape[0]
    for i in range(n_rows):
        if scipy.sparse.issparse(X):
            yield X.getrow(i)
        else:
            yield X[i]
```

Comment (JoanFM) [reply jina/drivers/encode.py:39]: Something like this, but we do not even need to changr the iterator, we can do the if else one level above (just be careful on how to check dependencies)

Comment (JoanFM) [reply jina/drivers/encode.py:39]: have a private `assign_embeddings` method that works differently for numpy arrays than for sparse arrays

File: tests/integration/sparse_pipeline/test_sparse_pipeline.py [hunk 2]
```diff
@@ -34,23 +34,23 @@ def encode(self, data: Any, *args, **kwargs) -> Any:
...
 
-
-    def query(
-        self, vectors: 'scipy.sparse.coo_matrix', top_k: int, *args, **kwargs
-    ):
+    def query(self, vectors: 'scipy.sparse.coo_matrix', top_k: int, *args, **kwargs):
         assert isinstance(vectors, sparse.coo_matrix)
         distances = [item for item in range(0, min(top_k, len(self.keys)))]
-        return np.array(self.keys[:top_k]), np.array(distances)
+        return np.array(self.keys[:top_k]), np.array(
```
Comment (davidbp) [inline tests/integration/sparse_pipeline/test_sparse_pipeline.py:51]: I think this was a bug, distances needs to be an array of arrays (for each vector I want the closest distances)

Comment (JoanFM) [reply tests/integration/sparse_pipeline/test_sparse_pipeline.py:51]: query is not done in batches

File: tests/integration/sparse_pipeline/test_sparse_pipeline.py [hunk 3]
```diff
@@ -66,10 +65,14 @@ def get_query_handler(self):
 
 def test_sparse_pipeline(mocker, docs_to_index):
     def validate(response):
-        print(response)
+        assert len(response.docs) == 10
```
Comment (JoanFM) [inline tests/integration/sparse_pipeline/test_sparse_pipeline.py:68]: we should assert the length of the matches, and the type of embedding returned in the matches

File: tests/integration/sparse_pipeline/test_sparse_pipeline.py [hunk 4]
```diff
@@ -66,10 +65,14 @@ def get_query_handler(self):
...
 
     f = Flow().add(uses=DummySparseEncoder).add(uses=DummyCSRSparseIndexer)
 
+    error_mock = mocker.Mock()
+
     with f:
         f.index(inputs=docs_to_index)
-        f.search(inputs=docs_to_index, on_done=validate)
+        f.search(inputs=docs_to_index, on_done=validate, on_error=error_mock)
```
Comment (JoanFM) [inline tests/integration/sparse_pipeline/test_sparse_pipeline.py:76]: use mock instead of directly validate, check other tests how they do

**Review by JoanFM — COMMENTED** [review]
We are missing unittests for new methods, for instancr as `all_sparse_embeddings` and some of the new sparse drivers

**Review by bwanglzu — APPROVED** [review]
LGTM!

**Pull request merged**